src/ray/util/pipe_logger.cc

```diff
@@ -191,43 +260,60 @@ RotationFileHandle CreatePipeAndStreamOutput(const std::string &fname,
 
   auto read_func = [read_fd](char *data, size_t len) { return Read(read_fd, data, len); };
   auto close_read_handle = [read_fd]() { RAY_CHECK_EQ(close(read_fd), 0); };
-  auto termination_caller = [write_fd]() {
+  auto close_fn = [write_fd, promise]() {
+    CompleteWriteEOFIndicator(write_fd);
     RAY_CHECK_EQ(close(write_fd), 0);
+    // Block until destruction finishes.
+    promise->get_future().get();
   };
 
-#elif defined(_WIN32)
-  SECURITY_ATTRIBUTES sa;
-  sa.nLength = sizeof(SECURITY_ATTRIBUTES);
-  sa.lpSecurityDescriptor = nullptr;
-  sa.bInheritHandle = TRUE;
-
-  HANDLE read_handle;
-  HANDLE write_handle;
-  RAY_CHECK(CreatePipe(&read_handle, &write_handle, &sa, 0));
+  auto logger = CreateLogger(log_redirect_opt);
 
-  auto read_func = [read_handle](char *data, size_t len) {
-    return Read(read_handle, data, len);
-  };
-  auto close_read_handle = [read_handle]() { RAY_CHECK(CloseHandle(read_handle)); };
-  auto termination_caller = [write_handle, read_handle]() {
-    RAY_CHECK(CloseHandle(write_handle));
+  // [content] doesn't have trailing newliner.
+  auto write_fn = [logger,
+                   log_redirect_opt = log_redirect_opt,
+                   std_stream_fd = std_stream_fd](const std::string &content) {
+    if (logger != nullptr) {
+      logger->log(spdlog::level::info, content);
+    }
+    if (log_redirect_opt.tee_to_stdout) {
+      RAY_CHECK_EQ(write(std_stream_fd.stdout_fd, content.data(), content.length()),
+                   content.length());
+      RAY_CHECK_EQ(write(std_stream_fd.stdout_fd, "\n", 1), 1);
+    }
+    if (log_redirect_opt.tee_to_stderr) {
+      RAY_CHECK_EQ(write(std_stream_fd.stderr_fd, content.data(), content.length()),
+                   content.length());
+      RAY_CHECK_EQ(write(std_stream_fd.stderr_fd, "\n", 1), 1);
+    }
   };
+  auto flush_fn =
+      [logger, log_redirect_opt = log_redirect_opt, std_stream_fd = std_stream_fd]() {
+        if (logger != nullptr) {
+          logger->flush();
+        }
+        if (log_redirect_opt.tee_to_stdout) {
+          fsync(std_stream_fd.stdout_fd);
+        }
```

Comment:
tee_to_stderr case?

Reply:
No need, stderr doesn't buffer anything.

Reply:
ok, can you add a comment saying that in the code